client/state/data-layer/wpcom/sites/comments/index.js

```diff
@@ -105,7 +111,12 @@ export const receiveCommentSuccess = ( action, response ) => {
 	return receiveComments( { siteId, postId, comments: [ response ], commentById: true } );
 };
 
-export const receiveCommentError = ( { siteId, commentId } ) => {
+export const receiveCommentError = ( { siteId, commentId, query = {} } ) => {
+	//if we see a force=wpcom, on retry try dropping the query param to see if the remote has the comment.
+	const { force, ...retryQuery } = query;
+	if ( force === 'wpcom' ) {
+		return requestCommentAction( { siteId, commentId, query: retryQuery } );
+	}
```

Comment:
This is a cool thing I think! It may be the first time I've seen the kind of fetching-specific error-handling logic being contained here in the data layer.

On that note I think it would be great if we could try to differentiate between network failures and shadow site failures. Is that even possible? Can we infer from the returned error (second parameter to this handler) what happened and only move over if we can judge that the sync had failed? I can imagine that we'll be in the dark either way, though if that's the case a good explanatory comment would be incredibly valuable here…

```js
// we can't tell the difference between a network failure and a shadow sync failure
// so if the request drops out automatically retry against the real site
```

It could be quite useful to know how often the shadow sites are missing information.

Reply:
> Can we infer from the returned error (second parameter to this handler) what happened and only move over if we can judge that the sync had failed?

Oh interesting thought. I'll def take your comment wording for the moment.

The most common case is that the comment is missing on the shadow site, so the error will be a 404 with "unknown_comment". I could add some meta on the retry maybe, so if we see a 404, we add meta, and if it's successful we can bump an analytics event to mark that a site is not in a good state. It's probably not useful per comment, but something like the date/site might be useful.

Reply:
Even detecting a 404 could be handy I think, though I don't believe that the data layer is currently exposing the status code; it should be returning headers.

If you think this would be helpful let me know or toss together some code. It seems like a very reasonable change we could make.